### PR TITLE
StoragePowerActor, StorageMinerActor et al.

### PR DESCRIPTION
**Pull request by jzimmerman**
This PR resolves most of the outstanding TODOs in the storage/mining-related actors, and introduces several updates, including:
- Cron dispatch via StoragePowerActor
- Sector/miner fault state machine isolation
- Improved event handling
- TemporaryFault logic
- Caller verification and associated machinery
- Exec for StorageMinerActors
- Funds checks for TemporaryFault declarations
- Power accounting improvements and fixes

### CODE REVIEW

File: src/systems/filecoin_vm/interpreter/vm_interpreter.go [hunk 1]
```diff
@@ -177,7 +178,9 @@ func (vmi *VMInterpreter_I) ApplyMessage(
 
 	// Check sender balance.
 	gasLimitCost := _gasToFIL(message.GasLimit(), message.GasPrice())
-	totalCost := message.Value() + actor.TokenAmount(gasLimitCost)
+	networkTxnFee := indices.Indices_FromStateTree(inTree).NetworkTransactionFee(
+		inTree.GetActorCodeID_Assert(message.To()), message.Method())
+	totalCost := message.Value() + gasLimitCost + networkTxnFee
```
Comment (zixuanzh) [inline src/systems/filecoin_vm/interpreter/vm_interpreter.go]: nice 👍

File: src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_state.go [hunk 2]
```diff
@@ -98,36 +99,104 @@ func (st *StoragePowerActorState_I) _selectMinersToSurprise(challengeCount int,
...
+	st._setClaimedPowerEntryInternal(minerAddr, currentPower-sectorPower)
+	st._updatePowerEntriesFromClaimedPower(minerAddr)
+}
+
+func (st *StoragePowerActorState_I) _updatePowerEntriesFromClaimedPower(minerAddr addr.Address) {
+	claimedPower, ok := st.ClaimedPower()[minerAddr]
+	Assert(ok)
+
+	nominalPower := claimedPower
+	if st.PoStDetectedFaultMiners()[minerAddr] {
```
Comment (zixuanzh) [inline src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_state.go]: will good to include a comment here on what nominal power is and what it is used for

File: src/systems/filecoin_vm/actor_util/actor_util.go [hunk 3]
```diff
@@ -158,3 +167,26 @@ func (x *DealIDQueue_I) Dequeue() (dealID deal.DealID, ok bool) {
 		return
 	}
 }
+
+func MinerSetHAMT_Empty() MinerSetHAMT {
```
[the remark this answers is not in the document]
Comment (jzimmerman) [reply src/systems/filecoin_vm/actor_util/actor_util.go]: Yes, could be -- not sure what persistent data structure will be most efficient for implementations, so leaving it in abstract form in the spec at the moment.

File: src/systems/filecoin_vm/actor_util/actor_util.id [hunk 4]
```diff
@@ -15,3 +17,25 @@ type DealIDQueue struct {
...
+
+type MinerEventSetHAMT {MinerEvent: bool}
+
+type SectorStorageWeightDesc struct {
+    SectorSize  sector.SectorSize
+    Duration    block.ChainEpoch
+    DealWeight  BigInt
+}
+
+type SectorTerminationType enum {
```
[the remark this answers is not in the document]
Comment (jzimmerman) [reply src/systems/filecoin_vm/actor_util/actor_util.id]: I don't think this is needed -- exceeding the limit of consecutive SurprisePoSt failures doesn't terminate all sectors individually, but rather terminates the miner as a whole. Adding a comment to address this.

File: src/systems/filecoin_mining/sector/sealing.id [hunk 5]
```diff
@@ -48,22 +48,22 @@ type OnChainSealVerifyInfo struct {
 type SealCommitment struct {
     SealedCID   SealedSectorCID  // CommR
     DealIDs     deal.DealIDs
-    Expiration  block.ChainEpoch
+    Expiration  block.ChainEpoch  // Note: may be zero, for pure capacity commitment with no guaranteed lifetime
```
[the remark this answers is not in the document]
Comment (jzimmerman) [reply src/systems/filecoin_mining/sector/sealing.id]: Yes, good catch -- this is out of date. Removing here and below.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id [hunk 6]
```diff
@@ -1,166 +1,99 @@
...
+
+    // SurprisePoSt challenge state: The miner has not submitted timely ElectionPoSts,
+    // and as a result, the system has fallen back to proving storage via SurprisePoSt.
+    Challenged struct {
+        SurpriseChallengeEpoch  block.ChainEpoch
+        ChallengedSectors       [sector.SectorNumber]
+        NumConsecutiveFailures  int
+    }
+
+    // SurprisePoSt faulted state: The miner has failed a SurprisePoSt, indicating that
```
[the remark this answers is not in the document]
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: The repeat is intentional -- both cases of the union need that information.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id [hunk 7]
```diff
@@ -1,166 +1,99 @@
...
+    DeclaredFaultEpoch          block.ChainEpoch  // -1 if not currently declared faulted.
+    DeclaredFaultDuration       block.ChainEpoch  // -1 if not currently declared faulted.
+    DealWeight                  BigInt  // -1 if not yet validated with StorageMarketActor.
+
+    // Must be significantly larger than DeclaredFaultEpoch, since otherwise it may be possible
+    // to declare faults adaptively in order to exempt challenged sectors.
+    EffectiveFaultBeginEpoch()  block.ChainEpoch
+    EffectiveFaultEndEpoch()    block.ChainEpoch
+
+    Is_TemporaryFault()         bool
```
[the remark this answers is not in the document]
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: I'm not sure either -- I think SectorOnChainInfo would be cleaner (and more efficient) as a union, but wanted to minimize the number of changes we're making at once.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id [hunk 8]
```diff
@@ -1,166 +1,99 @@
...
+    // Can be filled in once InvokeMethod is implemented for this actor.
 }
 
 type MinerInfo struct {
     // Account that owns this miner.
     // - Income and returned collateral are paid to this address.
     // - This address is also allowed to change the worker address for the miner.
-    Owner                   address.Address  // Must be an ID-address.
-    OwnerKey                filcrypto.VRFPublicKey
+    Owner                   address.Address
```
[the remark this answers is not in the document]
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: Ah, you're right; looks like that may have gotten lost in merge conflict resolution. I'll restore it.

File: src/systems/filecoin_mining/sector/sealing.id [hunk 9]
```diff
@@ -48,22 +48,22 @@ type OnChainSealVerifyInfo struct {
...
-    Expiration    block.ChainEpoch
+    Expiration    block.ChainEpoch  // Note: may be zero, for pure capacity commitment with no guaranteed lifetime
 }
 
 type SectorProveCommitInfo struct {
     SectorNumber
     Proof             SealProof
     InteractiveEpoch  block.ChainEpoch
-    Expiration        block.ChainEpoch
+    Expiration        block.ChainEpoch  // Note: may be zero, for pure capacity commitment with no guaranteed lifetime
```
[the remark this answers is not in the document]
Comment (jzimmerman) [reply src/systems/filecoin_mining/sector/sealing.id]: Agreed, I think many of the entries in these structs may be redundant, but wanted to stay as close as possible to the previous design. Let's plan to do a cleanup pass in a followup?

**Comment by sternhenri** [comment]
Note that as of this PR i no longer understand a lot of what is happening with storage miner actor logic and can no longer review its correctness. I believe a lot of this is complex beyond necessity and likely to cause bugs.

We should discuss this PR live.

**Comment by jzimmerman** [comment]
> We should discuss this PR live. A lot of great stuff, and some changes that make the actor code more opaque. Would love to go through it sync.

Sounds good. @zixuanzh and I discussed in detail before the holidays and we tried to keep the logic as streamlined as possible, but some additional complexity was needed to implement the desired mechanisms for independent sector faulting and graceful recovery. Happy to discuss sync as well.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id [hunk 10]
```diff
@@ -1,166 +1,99 @@
...
+
+    Is_TemporaryFault()         bool
 }
 
-type PreCommittedSectorsAMT {sector.SectorNumber: PreCommittedSector}
-type StagedCommittedSectorAMT {sector.SectorNumber: SectorOnChainInfo}
 type SectorsAMT {sector.SectorNumber: SectorOnChainInfo}
 
-// Balance of a StorageMinerActor should equal exactly the sum of PreCommit deposits that are not yet returned or burned
+type SectorNumberSetHAMT {sector.SectorNumber: bool}
```
Comment (anorth) [inline src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: I think we developed the RLE+ encoding specifically to represent sets of sector numbers efficiently.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id [hunk 11]
```diff
@@ -1,166 +1,99 @@
...
     // - Income and returned collateral are paid to this address.
     // - This address is also allowed to change the worker address for the miner.
     Owner                   address.Address  // Must be an ID-address.
-    OwnerKey                filcrypto.VRFPublicKey
 
     // Worker account for this miner.
     // This will be the key that is used to sign blocks created by this miner, and
     // sign messages sent on behalf of this miner to commit sectors, submit PoSts, and
     // other day to day miner activities.
-    Worker                  address.Address  // Must be an ID-address.
```
Comment (anorth) [inline src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: Why did you remove this constraint? Note that the runtime immediate caller will always present an ID-address, so caller validation will fail unless this address matches.

Note that the key can be looked up from the address so the field below is unnecessary.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go [hunk 12]
```diff
@@ -3,106 +3,150 @@ package storage_mining
...
-	rt.ValidateImmediateCallerIs(addr.CronActorAddr)
-
-	a._checkSurprisePoStSubmissionHappened(rt)
-
-	return rt.SuccessReturn()
+func (a *StorageMinerActorCode_I) GetWorkerVRFKey(rt Runtime) filcrypto.VRFPublicKey {
+	h, st := a.State(rt)
+	ret := st.Info().WorkerVRFKey()
+	Release(rt, h, st)
+	return ret
```
Comment (anorth) [inline src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go]: Why does this exist? It is never invoked.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go [hunk 13]
```diff
@@ -3,106 +3,150 @@ package storage_mining
...
-func (a *StorageMinerActorCode_I) PreCommitSector(rt Runtime, info sector.SectorPreCommitInfo) InvocOutput {
-	h, st := a.State(rt)
+//////////////////
+// ElectionPoSt //
+//////////////////
 
-	rt.ValidateImmediateCallerIs(st.Info().Worker())
-	// can be called regardless of Challenged status
+// Called by the VM interpreter once an ElectionPoSt has been verified.
+// Assumes the block reward has already been granted to the storage miner actor.
```
Comment (anorth) [inline src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go]: We intent to grant block reward at the end of the block, so this will be false. But I can't see anything below that depends on it anyway.

File: src/systems/filecoin_vm/interpreter/vm_interpreter.go (hunk 1 above)
Comment (anorth) [reply src/systems/filecoin_vm/interpreter/vm_interpreter.go]: I don't think this is good. Even if there was some additional fee as part of sending a message, it still has to fit within the gas limit so a sender can bound their cost.

File: src/systems/filecoin_vm/runtime/impl/runtime.go [hunk 14]
```diff
@@ -191,6 +193,35 @@ func (rt *VMContext) _createActor(codeID actor.CodeID, address addr.Address) {
...
+	}
+
+	// Special case: StoragePowerActor may delete a StorageMinerActor.
+	addrCodeID, found := rt.GetActorCodeID(address)
+	if found &&
+		rt._actorAddress.Equals(addr.StoragePowerActorAddr) &&
+		addrCodeID.Is_Builtin() && addrCodeID.As_Builtin() == actor.BuiltinActorID_StorageMiner {
+
+		ok = true
+	}
```
Comment (anorth) [inline src/systems/filecoin_vm/runtime/impl/runtime.go]: This coupling is totally unnecessary. Just have the SPA send a DeleteSelf() message to the SMA.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id (hunk 10 above)
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: This is currently used for the ProvingSet, and it's not clear RLE+ gives a more efficient encoding in that case (especially as sector numbers continue to be incremented over time).

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id (hunk 11 above)
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: You're right, this got lost in a merge conflict resolution -- I'll restore it in a new PR.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go (hunk 13 above)
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go]: Yes, these comments are outdated; I'll remove them in a subsequent PR.

File: src/systems/filecoin_vm/interpreter/vm_interpreter.go (hunk 1 above)
Comment (jzimmerman) [reply src/systems/filecoin_vm/interpreter/vm_interpreter.go]: > it still has to fit within the gas limit so a sender can bound their cost

I don't see how that follows. I think the sender can bound their cost even if it doesn't fit within the gas limit, the bound just has to be increased to include the maximum transaction fee for the given method.

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go (hunk 13 above)
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go]: #806

File: src/systems/filecoin_vm/runtime/impl/runtime.go (hunk 14 above)
Comment (jzimmerman) [reply src/systems/filecoin_vm/runtime/impl/runtime.go]: #806

File: src/systems/filecoin_mining/storage_mining/storage_miner_actor.id (hunk 11 above)
Comment (jzimmerman) [reply src/systems/filecoin_mining/storage_mining/storage_miner_actor.id]: #806
